tools/perf/ucc_pt_bootstrap.h

```diff
@@ -8,14 +8,66 @@
 #define UCC_PT_BOOTSTRAP_H
 
 #include <ucc/api/ucc.h>
+#include <unistd.h>
+#include <functional>
+#include <string>
+#include <iostream>
 
 class ucc_pt_bootstrap {
 protected:
+    size_t node_hash;
     ucc_context_oob_coll_t context_oob;
     ucc_team_oob_coll_t team_oob;
+    int ppn;
+    int find_ppn()
+    {
+        int     comm_size = get_size();
+        size_t *hashes    = new size_t[comm_size];
+        int     ppn       = 0;
+        ucc_status_t st;
+        void *req;
+
+        st = team_oob.allgather(&node_hash, hashes, sizeof(node_hash),
+                                                    team_oob.coll_info, &req);
+        if (st != UCC_OK) {
+            goto exit_err;
+        }
+        do {
+            st = team_oob.req_test(req);
+        } while (st == UCC_INPROGRESS);
+        if (st != UCC_OK) {
+            goto exit_err;
+        }
+        team_oob.req_free(req);
+        for (int i = 0; i < comm_size; i++) {
+            if (node_hash == hashes[i]) {
+                ppn++;
+            }
+        }
+        delete[] hashes;
+        return ppn;
+exit_err:
+        std::cerr <<"failed to find ppn" <<std::endl;
+        delete[] hashes;
+        return ppn;
+    }
 public:
+    ucc_pt_bootstrap()
+    {
+        char hostname[256];
+        gethostname(hostname, sizeof(hostname));
+        node_hash = std::hash<std::string>{}(std::string(hostname));
```

Comment:
I've got a question regarding node_hash in general. I know we used that in several places, but what about the collision ? It can happen right, and there is no way even to detect it. The  code will not work and it is even hard to understand why. What should we do (asking for future, for UCC hierarchy discovery) ?
In this case btw, you can make get_ppn virtual and in MPI case you can use OMPI/MPICH env vars or fallback to base class if no env var found.

Reply:
perhaps we can use ucs_machine_guid. It returns uint64_t based on mac address and host name, but it's not public currently